ThirdPartyModules/Magento/GiftCardAccount.php

```diff
@@ -196,4 +196,127 @@ function boltMagentoGiftInvalidateCart() {
+        $code,
+        $giftCard,
+        $immutableQuote,
+        $parentQuote
+    ) {
+        if (!$giftCard instanceof \Magento\GiftCardAccount\Model\Giftcardaccount) {
+            return $result;
+        }
+        try {
+            if ($immutableQuote->getGiftCardsAmountUsed() == 0) {
+                try {
+                    // on subsequest validation calls from Bolt checkout
+                    // try removing the gift card before adding it
+                    $giftCard->removeFromCart(true, $immutableQuote);
+                } catch (\Exception $e) {
+                    // gift card not added yet
+                } finally {
+                    $giftCard->addToCart(true, $immutableQuote);
+                }
+            }
+
+            if ($parentQuote->getGiftCardsAmountUsed() == 0) {
+                try {
+                    // on subsequest validation calls from Bolt checkout
+                    // try removing the gift card before adding it
+                    $giftCard->removeFromCart(true, $parentQuote);
+                } catch (\Exception $e) {
+                    // gift card not added yet
+                } finally {
+                    $giftCard->addToCart(true, $parentQuote);
+                }
+            }
+
+            // Send the whole GiftCard Amount.
+            $giftAmount = $parentQuote->getGiftCardsAmount();
+            $currencyCode = $parentQuote->getQuoteCurrencyCode();
+
+            return [
+                'status'          => 'success',
+                'discount_code'   => $code,
+                'discount_amount' => abs(
+                    CurrencyUtils::toMinor($giftAmount, $currencyCode)
+                ),
+                'description'     => __('Gift Card (%1)', $code),
+                'discount_type'   => 'fixed_amount',
+            ];
+        } catch (\Exception $e) {
+            return [
+                'status'        => 'failure',
+                'error_message' => $e->getMessage(),
+            ];
+        }
+    }
+    
+    /**
+     * @param bool                                           $result
+     * @param string                                         $couponCode
+     * @param \Magento\GiftCardAccount\Model\Giftcardaccount $giftCard
+     * @param Quote                                          $quote
+     * @return bool
+     */
+    public function filterApplyingGiftCardCode(
+        $result,
+        $couponCode,
+        $giftCard,
+        $quote
+    ) {
+        if ($result || !$giftCard instanceof \Magento\GiftCardAccount\Model\Giftcardaccount) {
+            return $result;
+        }
+        
+        try {
+            // on subsequest validation calls from Bolt checkout
+            // try removing the gift card before adding it
+            $giftCard->removeFromCart(true, $quote);
+        } catch (\Exception $e) {
+            // gift card not added yet
+        } finally {
+            $giftCard->addToCart(true, $quote);
```

Comment:
Should we try / catch / return false here as well?

Reply:
@danac-gs In this PR, we just refactor the existing codes. and for your concern, I do notice it as well, and shall we move to this task https://boltpay.atlassian.net/browse/M2P-468 for discussion?

Reply:
I was just noticing that, comparing to the below `filterRemovingGiftCardCode` method, one of then can throw and another one can't, and they should probably act consistently. Not sure what the right approach is, but the change is small and could be implemented here if we agree. cc @vitaliyreznikov